apps/api/src/app/events/helpers/helpers.ts

```diff
@@ -0,0 +1,94 @@
+import { StepTypeEnum, DelayTypeEnum, DigestUnitEnum, DigestTypeEnum } from '@novu/shared';
+import { sub, differenceInMilliseconds } from 'date-fns';
+import { ApiException } from '../../shared/exceptions/api.exception';
+import { NotificationStepEntity } from '@novu/dal';
+import { get } from 'lodash';
+
+export type StepWithDelay = NotificationStepEntity & { delay: number };
+
+//digest and delay with delayPath always on top of dag, delays are added to next active node and removed delay nodes
+export function constructActiveDAG(steps: NotificationStepEntity[], payload = {}, overrides = {}): StepWithDelay[][] {
+  const dag = constructBasicDAG(steps);
+
+  return overrideDelaysAndRemoveDelaySteps(dag, payload, overrides);
+}
+
+function constructBasicDAG(steps: NotificationStepEntity[]): StepWithDelay[][] {
+  const resultDag = steps.reduce((dag, step, i) => {
+    if (step.template?.type === StepTypeEnum.DIGEST || step.metadata?.delayPath || dag.length == 0) dag.push([]);
+    dag[dag.length - 1].push({ ...step, delay: getMetaDelay(step) });
+
+    return dag;
+  }, [] as StepWithDelay[][]);
+
+  return resultDag.map((branch) => {
+    const inactiveIndex = branch.findIndex((step) => !step.active);
+
+    return inactiveIndex === -1 ? branch : branch.slice(0, inactiveIndex);
+  });
```

Comment:
It is hard to follow up the code and I don't even know what DAG stands for (acronyms without any context nor documentation are not very helpful unfortunately) but I guess this is the place where the new main logic starts, is that right? 
My surface understanding is that here you create it like a list in order of the different jobs a triggered event needs to execute. If that's right, it is an idea we were considering to solve some of the problems we were having: to create all the jobs needed to execute before queuing and then change the status or cancel them based on the other functionalities we have before executing them from the queue. That would help to remove many calculations and the locks we had to introduce to assure right behaviour against concurrency (at least for the regular Digest; Backoff digest still not working properly; I neither found here that you fully fixed it).

One of the problems with this implementation is to consider if in the future we want to implement conditional workflows (think in decision trees) how flexible a linear creation before queueing would be. So that would need time to think about it.

Reply:
DAG means Directed Acyclic Graph, to make it method name short I used DAG, coming to functionality we are taking the linear chain and constructing a gragh with the parallel paths, lets say if we have SMS1->Email1->Digest1->Email2->Digest2-Chat1 then it will create a parallel paths: 
1. SMS1->Email1
2. Digest1-> Email2
3. Digest2 -> Chat1

All three paths can executes parallel, and coming to jobs creation, no need to create upfront, that is our wish, when ever needed we can construct these DAGs, if needed we can cache the basic DAG along with template(when we have big chains caching may help), please look at spec tests file https://github.com/tsssdev/novu/blob/d058a888b28a144bb46956f4eb9847d772d553c3/apps/api/src/app/events/helpers/helper.dag.spec.ts for better understanding.
Coming to use cases, what ever use cases I can think off, the new code can handle very easily.. here are few use cases:
1. Multiple digest nodes(this is already implemented in new code)
2. Parallel jobs, instead of one sequential chain (already support is builtin in new code)
3. Daily/weekly digests at specified time, easy
4. deliver messages at the subscriber time zone.. in this new code easy
5. Count based digest, instead of time based if they want to deliver for every 1000(batch), it is very easy to implement in new code..
6. If they want to mix with count and time-based digest(which ever comes first), again easy in new code
7. Backoff after sending 10 messages, again easy.
8. Backoff count and timebased, easy.
9. cron/rrule support, easy

Throw me a use case you think tough then I can try to show case how easy to implement in this new code.

One more thing is even though I started with digest refactoring, I refactored complete jobs code, so not only digest jobs, even regular jobs benefit with these changes. and delay and trigger jobs overhead is gone(functionality not effected)